Extended colors should be clamped so they look the same in SDR and HDR layers.

mattwoodrow opened the pull request:
#### fd22b8420d86a73fc777a825132d29f1ebb3b85d
<pre>
Extended colors should be clamped so they look the same in SDR and HDR layers.
<a href="https://bugs.webkit.org/show_bug.cgi?id=292226">https://bugs.webkit.org/show_bug.cgi?id=292226</a>
&lt;<a href="https://rdar.apple.com/150231788">rdar://150231788</a>&gt;

Reviewed by Mike Wyrzykowski.

When drawing to an extended color space destination, manually convert colors
into the standard range to get the same output as would have happened with a
standard range destination.

* Source/WebCore/platform/graphics/Color.h:
* Source/WebCore/platform/graphics/cg/ColorCG.cpp:
(WTF::RetainPtr&lt;CGColorRef&gt;&gt;::createValueForKey):
(WebCore::createSDRCGColorForColorspace):
(WebCore::cachedSDRCGColorForColorspace):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGFillColor):
(WebCore::GraphicsContextCG::drawRect):
(WebCore::GraphicsContextCG::drawLine):
(WebCore::GraphicsContextCG::fillRect):
(WebCore::GraphicsContextCG::fillRoundedRectImpl):
(WebCore::GraphicsContextCG::setCGShadow):
(WebCore::GraphicsContextCG::didUpdateState):
(WebCore::GraphicsContextCG::drawLinesForText):
* Tools/TestWebKitAPI/Tests/WebCore/cg/GraphicsContextCGTests.mm:
(TestWebKitAPI::TEST(GraphicsContextTests, OutOfGamutSRGBNotDrawn)):

Canonical link: <a href="https://commits.webkit.org/297279@main">https://commits.webkit.org/297279@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6143be815849cbf391fd08150a60cf6389e866aa

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/110851 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/30514 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/20948 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/116880 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61119 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/112814 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31194 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39100 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84276 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/113799 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/24919 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/99809 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/64720 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24275 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/17951 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/60677 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94303 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18012 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/119696 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/37896 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28158 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93248 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38272 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96083 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93073 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23764 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38107 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/15855 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/33870 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/37789 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43259 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/37451 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/40788 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39157 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

weinig commented on Source/WebCore/platform/graphics/cg/ColorCG.cpp:
```diff
@@ -212,6 +212,20 @@ RetainPtr<CGColorRef> cachedCGColor(const Color& color)
     return cache.get().get(color);
 }
 
+RetainPtr<CGColorRef> cachedCGColorFor(const Color& color, CGColorSpaceRef colorSpace)
```
Given that this function does something non-trivial, I think it needs a more clear name to indicate it has this behavior for extended range color spaces.

mattwoodrow replied:
Also agreed :)

I'll try think of something better than `cachedCGColorMaybeMappedIntoSDRRangeFor` (and add a settings check for this behaviour).

weinig reviewed (changes requested):
There are few things that would help make reviewing this easier (I would check, but I don't have easy access to a Tahoe beta or any iOS right now):

- What are the different values of `colorSpace()` you see being returned from `GraphicsContext::colorSpace()` currently? 
- What is it for a window on a display without any HDR capability vs. one with HDR capability? 
- Does it change when there if there is HDR content being displayed, or does it always exactly match the output display.

Beyond that, I think you will likely want to implement the following optimizations:

1. In `cachedCGColorFor`, if the input color is trivially within SDR range (for instance, you could check if is stored as SRGBA<uint8_t> in Color or if it has the same color space but all its components are between 0.0 and 1.0), you should avoid any additionally work. This will be the vast majority of colors.
2. You should cache the "standard range" color space for a context on the context, rather than recomputing on each call to `cachedCGColorFor`.
3. Avoid multiple color conversions by refactoring cachedCGColor() to take an optional color space, and making the TinyLRUCache be keyed on the pair of (color, desired final colorSpace).

mattwoodrow committed:
Extended colors should be clamped so they look the same in SDR and HDR layers.
https://bugs.webkit.org/show_bug.cgi?id=292226
<rdar://150231788>

Reviewed by Mike Wyrzykowski.

When drawing to an extended color space destination, manually convert colors
into the standard range to get the same output as would have happened with a
standard range destination.

* Source/WebCore/platform/graphics/Color.h:
* Source/WebCore/platform/graphics/cg/ColorCG.cpp:
(WTF::RetainPtr<CGColorRef>>::createValueForKey):
(WebCore::createSDRCGColorForColorspace):
(WebCore::cachedSDRCGColorForColorspace):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::setCGFillColor):
(WebCore::GraphicsContextCG::drawRect):
(WebCore::GraphicsContextCG::drawLine):
(WebCore::GraphicsContextCG::fillRect):
(WebCore::GraphicsContextCG::fillRoundedRectImpl):
(WebCore::GraphicsContextCG::setCGShadow):
(WebCore::GraphicsContextCG::didUpdateState):
(WebCore::GraphicsContextCG::drawLinesForText):
* Tools/TestWebKitAPI/Tests/WebCore/cg/GraphicsContextCGTests.mm:
(TestWebKitAPI::TEST(GraphicsContextTests, OutOfGamutSRGBNotDrawn)):

Canonical link: https://commits.webkit.org/297279@main
